daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/FNFunctions.scala

```diff
@@ -803,6 +803,52 @@ case class FNFloor(recipe: CompiledDPath, argType: NodeInfo.Kind) extends FNOneA
   }
 }
 
+case class DFDLXLeftShift(recipes: List[CompiledDPath],argType: NodeInfo.Kind) extends FNTwoArgs(recipes) {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+
+    val shift = asInt(arg2.getAnyRef).intValue
+    argType match {
+      case NodeInfo.Long => {
+        val leftShift = asLong(arg1).longValue()
+        if (shift >=64) {
+          0
+        }
+        else {
+          leftShift << shift
+        }
+      }
+      case NodeInfo.Int => {
+        val leftShift = asInt(arg1.getAnyRef).intValue()
+        if (shift >=32) {
+           0
+        }
+        else {
+           leftShift << shift
+        }
+      }
+      case NodeInfo.Short => {
+        val leftShift = asShort(arg1.getAnyRef).shortValue()
+        if (shift >=16) {
+           0
+        }
+        else {
+           leftShift << shift
+        }
+      }
+      case NodeInfo.Byte => {
+        val leftShift = asByte(arg1.getAnyRef).byteValue()
+        if (shift >=8) {
+           0
+        }
+        else {
+           leftShift << shift
+        }
+      }
```

Comment:
Note that we have ULong, UInt, UShort and UByte classes which are used to represent unsigned integers and those already have shift operations defined on them, so no masking etc. needs to be figured out.  These are in the passera.unsigned package in daffodil-lib.

Reply:
Note that the DataValuePrimitive stuff doesn't use any of the Unsigned* types we have. It uses all Java pimiritves (or their Boxed versions). So we can potentially use those, but ultilately the data value needs to be converted back to the correct java primitive.

Reply:
It should be as simple as:
```
ULong(dv.getLong()) << count // shift left
ULong(dv.getLong()) >> count // logical right shift
```

Reply:
I have converted  short, byte,UnsignedByte back.Rest all will automatically will converted to int and long.
For UnsignedByte also, I used toByte because for java everything is signed.
@stevedlawrence let me know if this works

Reply:
Missing case for unsigned long.